### PR TITLE
[Hybrid parallel] Add sequence parallel

### PR DESCRIPTION
**Pull request by DesmonDay**
### PR types
New features

### PR changes
Others

### Description
Add sequence parallel method for Llama model.

### CODE REVIEW

File: paddlenlp/transformers/llama/configuration.py [hunk 1]
```diff
@@ -211,6 +211,8 @@ def __init__(
         use_flash_attention=False,
         use_fused_rms_norm=False,
         tensor_parallel_output=True,
+        sequence_parallel=False,
+        seq_len=1024,
```
Comment (ZHUI) [inline paddlenlp/transformers/llama/configuration.py]: 一定要 seq_len？

File: paddlenlp/transformers/llama/modeling.py [hunk 2]
```diff
@@ -421,23 +453,28 @@ def forward(
         use_cache: bool = False,
     ) -> Tuple[paddle.Tensor, Optional[paddle.Tensor], Optional[Tuple[paddle.Tensor]]]:
         """Input shape: Batch x Time x Channel"""
-        bsz, q_len, _ = hidden_states.shape
-        query_states = self.q_proj(hidden_states).reshape(shape=[bsz, q_len, self.num_heads, self.head_dim])
-        key_states = self.k_proj(hidden_states).reshape(shape=[bsz, q_len, self.num_heads, self.head_dim])
-        value_states = self.v_proj(hidden_states).reshape(shape=[bsz, q_len, self.num_heads, self.head_dim])
+        # [bs, seq_len, num_head * head_dim] -> [seq_len / n, bs, num_head * head_dim] (n is model parallelism)
+        query_states = self.q_proj(hidden_states).reshape(shape=[0, 0, self.num_heads, self.head_dim])
```
Comment (ZHUI) [inline paddlenlp/transformers/llama/modeling.py]: 就 保留 `bsz, seq_len` 的写法吧，0,0 其实很不直观。傍边注释一下，sequence_parallel 下，其实是 `seq_len， bsz`

File: paddlenlp/transformers/llama/modeling.py [hunk 3]
```diff
@@ -452,6 +489,12 @@ def forward(
             attention_mask=attention_mask,
             output_attentions=output_attentions,
         )
+
+        if self.sequence_parallel:
```
Comment (ZHUI) [inline paddlenlp/transformers/llama/modeling.py:604]: 你这个直接放最后吧 `attn_output = self.o_proj(attn_output)` 之后

File: paddlenlp/transformers/llama/modeling.py [hunk 4]
```diff
@@ -887,6 +944,10 @@ def __init__(self, config: LlamaConfig):
             self.weight.split_axis = 1
 
     def forward(self, hidden_states, tensor_parallel_output=None):
+        if self.config.sequence_parallel:
+            hidden_states = GatherOp.apply(hidden_states)
+            hidden_states = hidden_states.reshape([-1, self.config.seq_len, hidden_states.shape[-1]])
```
Comment (ZHUI) [inline paddlenlp/transformers/llama/modeling.py]: 一定要弄个参数？

File: paddlenlp/transformers/llama/modeling.py (hunk 3 above)
Comment (DesmonDay) [reply paddlenlp/transformers/llama/modeling.py:604]: 不太行，对于 sequence_parallel 情况下，o_proj 的输入 shape 需要是 [seq_len, bs, dim]，所以需要提前 reshape。

File: paddlenlp/transformers/llama/configuration.py (hunk 1 above)
Comment (DesmonDay) [reply paddlenlp/transformers/llama/configuration.py]: 去掉了

File: paddlenlp/transformers/llama/modeling.py (hunk 4 above)
Comment (DesmonDay) [reply paddlenlp/transformers/llama/modeling.py]: 去掉了

File: paddlenlp/trainer/training_args.py [hunk 5]
```diff
@@ -535,6 +539,22 @@ class TrainingArguments:
...
+        metadata={
+            "help": (
+                "Sequence parallelism is a parallel technique proposed in (https://arxiv.org/pdf/2205.05198.pdf see 4.2.2 Sequence Parallelism)."
+                "This technique partitions the non-tensor parallel regions of a transformer layer, whose operations are independent along the sequence dimension."
+                "Notes, this can only be True when model_parallel_world_size is larger than 1."
+            )
+        },
+    )
+
+    fuse_sequence_parallel_allreduce: bool = field(
```
Comment (ZHUI) [inline paddlenlp/trainer/training_args.py]: 如昨天沟通，不要这里加这些参数了

File: paddlenlp/transformers/llama/modeling.py (hunk 2 above)
Comment (DesmonDay) [reply paddlenlp/transformers/llama/modeling.py]: 如果要保留原来的写法，sp的需要另外计算出来reshape，没法直接用inputs的shape。

File: paddlenlp/transformers/llama/modeling.py (hunk 3 above)
Comment (ZHUI) [reply paddlenlp/transformers/llama/modeling.py:604]: 你知道最后 一维 做 matmu， 不是一样的？

File: llm/llama/run_trainer_tp4pp2sp.sh [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    --eval_steps 1000 \
+    --report_to "visualdl" \
+    --sharding "stage1" \
+    --disable_tqdm true \
+    --continue_training 1\
+    --recompute 1 \
+    --do_train \
+    --do_eval \
+    --device "gpu" \
+    --sequence_parallel
```
Comment (ZHUI) [inline llm/llama/run_trainer_tp4pp2sp.sh]: 不要新增文件了。原来的脚本里面加 `--sequence_parallel` 默认值False

File: llm/llama/run_pretrain.py [hunk 7]
```diff
@@ -434,6 +444,11 @@ def main():
     else:
         model = model_class._from_config(config, dtype=dtype)
 
+    if model_args.sequence_parallel:
+        register_sequence_parallel_allreduce_hooks(
```
Comment (ZHUI) [inline llm/llama/run_pretrain.py:464]: 看起来这些东西还得想办法集成到框架。

**Review by ZHUI — DISMISSED** [review]
LGTM

**Review by ZHUI — DISMISSED** [review]
LGTM

**Review by ZHUI — DISMISSED** [review]
LGTM

**Review by ZHUI — DISMISSED** [review]
